File: 2. Structural Design Patterns/Proxy/Go/ProxyTemplate.go
package main

// Proxy Pattern – Go Template

type Subject interface{ Request() }

type RealSubject struct{}
func (RealSubject) Request(){ /* heavy work */ }

type Proxy struct{ real *RealSubject }
func (p *Proxy) Request(){ if p.real == nil { p.real = &RealSubject{} }; /* checks/logging */ p.real.Request() }

func main(){ var s Subject = &Proxy{}; s.Request() }

File: 2. Structural Design Patterns/Proxy/Typescript/ProxyTemplate.ts
// Proxy Pattern – TypeScript Template

interface Subject { request(): void }
class RealSubject implements Subject { request(): void { /* heavy work */ } }
class Proxy implements Subject { private real: RealSubject | null = null; request(): void { if (!this.real) this.real = new RealSubject(); /* checks/logging */ this.real.request() } }

;(function main(){ const s: Subject = new Proxy(); s.request() })()

export {}
